### PR TITLE
"spack logs": print log files for packages (either partially built or installed)

### PR DESCRIPTION
**Pull request by scheibelp**
`spack log <spec>`:

Prints the log file to the terminal, automatically decompressing it if needed.

* If the package is installed, this uses the location of the log in the install directory
* If the package is not installed but staged, this uses the location of the log in the stage directory
* If the package is neither installed nor staged, this prints an error

(punted to later PR) ~For an in-progress build, you can do `spack log --continuous`, which will operate like `tail -f` on the log file as it is written to.~

(removed from this PR) ~`spack location --logs <spec>`: Prints the location of the file (same semantics as far as locating the log file).~

Notes: 

* (Update Jan 22) https://github.com/spack/spack/pull/42122 was merged, so I think in general providing a `spack location --logs` option is counterproductive, e.g. if you say `` grep -i "foo" `spack location --logs <pkg-name>` `` it would be searching for "foo" in the compressed file
* (no longer relevant as of Jan. 22: there is no more `--logs` option) ~`spack location --logs` is something of a departure from other `location` options, which tend to require that the user be explicit (e.g. if the user does `spack location foo`  the default is to print the stage directory of `foo`, and to print an error if it is not staged (even if it is installed).~
* (no longer relevant as of Jan. 22: ~there is no more `--logs` option) It was pointed out that also unlike other `spack location` options, `--logs` prints a file location vs. a directory, which adds some confusion about what to do. It was suggested that `setup-env.sh` could automatically extract the base directory in its defined bash function~

TODOs

- [x] Test `spack log`
- [ ] `spack log` would ideally be able to print an in-progress log file like `tail -f`
  - ~(update Jan. 24) A probably-functional version of this is now available and can be used by invoking `spack log --continuous`.~
  - (update Jan. 25) it turns out if we want to stream ongoing builds, we'll need to detect the individual phase logs and print each one as it shows up (or change `log_output` as suggested in https://github.com/spack/spack/pull/42202#discussion_r1467332882): the singular `spack-build-out.txt` is assembled after the Spack installation completes (whether successful or not)
- [x]  (resolved by removal of `--logs` option) `spack location --logs` can print paths that don't exist; e.g. if a user manually deletes a log file in the install directory, this doesn't check that it exists; also for example if the package has been staged but no build commands have been run, then it might be that no log content has yet been generated (and the log in the stage wouldn't exist), but this command would still print where the log *would* be once that occurred. Overall though, if the log path doesn't exist, it can at least be said that it doesn't exist elsewhere either.

### CODE REVIEW

**Comment by haampie** [comment]
What's the use case for printing the location of a log? Can't you just have a command dump build logs to stdout? And if that exists there is no point in a command that dumps the location, since users could `spack log ... > build.log` / `spack log -o build.log ...`).

Also note that installed logs are now compressed (#42122) so you have to look for both `*.txt` and `*.txt.gz` in the install prefix, which makes a command like ^ extra convenient cause users wouldn't have to conditionally decompress.

**Comment by scheibelp** [comment]
I've added `spack log` to this

I agree `spack location --logs` should be removed because it could actually mislead users when they feed the output to `grep` (explained further in the PR description).

I need to add tests for `spack log` (and also add support to simulate `tail -f` on in-progress logs). I'll leave the `--logs` option for now and have also marked this WIP

File: lib/spack/spack/cmd/log.py
Comment (scheibelp) [inline lib/spack/spack/cmd/log.py]: This appears to work for me on Mac OS, but I can't find any documentation that guarantees that it would (e.g. on Linux, I can't find conditions where an EOF indicator is unset other than when the user explicitly calls a function).

Comment (scheibelp) [reply lib/spack/spack/cmd/log.py]: I can find multiple threads which suggest this like https://stackoverflow.com/questions/1703640/how-to-implement-a-pythonic-equivalent-of-tail-f, but nothing definitive in the POSIX standard (https://pubs.opengroup.org/onlinepubs/9699919799/) which guarantees this ought to work (and it's also unknown to me whether this would work on Windows).

I don't think it's harmful to include it, and if we encounter any problems then it would be possible to write something higher-level (e.g. to monitor the file status for updates and reopen it if changes are detected). Without something like this, users don't have a means to `tail -f` ongoing builds.

File: lib/spack/spack/cmd/log.py
Comment (haampie) [inline lib/spack/spack/cmd/log.py]: Follow standards: `-f, --follow`, `--continuous` is not following the principle of least surprise.

File: lib/spack/spack/cmd/log.py
Comment (haampie) [reply lib/spack/spack/cmd/log.py]: I would honestly just execve tail -f on systems that support it, instead of reinventing the wheel. Or at least prefer that when available, and fall back to the poor man's sleep version, which may hit all the bugs already addressed over the decades in coreutils.

File: lib/spack/spack/cmd/log.py
Comment (haampie) [inline lib/spack/spack/cmd/log.py]: `spec.stage.log_path` can never be streamed, it's concatenated from individual phase logs after all phases are done. If you insist on this functionality then please make the log writer output to phase and combined log as it goes, cause you can't know when a phase finishes.

File: lib/spack/spack/cmd/log.py
Comment (haampie) [inline lib/spack/spack/cmd/log.py]: Use `gzip.open` to decorate the stream instead of using a temporary directory + chdir + temporary uncompressed file (+ decoding unicode). There's no point in generalizing here if the current implementation to write logs is not generic. (And I don't see why it should be, cause gzip achieves compression to ~4% of original size at decent speed)

It can just be

```python
def maybe_decompress_text(path: str) -> io.TextIOWrapper:
    return gzip.open(path, "rt") if log_path.endswith(".gz") else open(path)

with maybe_decompress_text(log_path) as f:
    shutil.copyfileobj(f, sys.stdout)
```

or better, open in binary mode (but notice you have to reopen `sys.stdout` in binary mode):

```python
def maybe_decompress(path: str):
    return gzip.open(path, "r") if log_path.endswith(".gz") else open(path, "rb")

def binary_stdout():
    return os.fdopen(sys.stdout.fileno(), "wb", closefd=False)

with maybe_decompress(log_path) as f, binary_stdout() as g:
    shutil.copyfileobj(f, g)
```

`spack.util.compression._py_gunzip` extracts a file in the current working directory? That's poor design :(

File: lib/spack/spack/cmd/log.py
Comment (haampie) [inline lib/spack/spack/cmd/log.py]: This logic is (confusingly?) part of `matching_spec_from_env`, no need to duplicate.

File: lib/spack/spack/cmd/log.py
Comment (haampie) [reply lib/spack/spack/cmd/log.py]: Also early return missing?

File: lib/spack/spack/cmd/log.py
Comment (haampie) [inline lib/spack/spack/cmd/log.py]: I would open in binary mode so you don't have to worry about potentially invalid unicode in logs, which happens every now and then.

**Review by haampie — CHANGES_REQUESTED** [review]
General request to keep it simple (open stream, optionally wrap in de-compressor, copy into stdout).

The current implementation pulls in tons of unnecessary things: temporary dir, chdir, having to provide a name for a temporary file, UTF-8 decoding issues, slowness because of double pass and reading line by line and even .strip()'ing those causing further allocations, ...)

Should just really be equivalent to `[ "${f##*.}" = "gz" ] && gzip -d < "$f" || cat "$f"`.

**Comment by haampie** [comment]
Also maybe worth looking at other tools and how they name this command, for example `docker logs` is plural, `kubectl logs` too.

File: lib/spack/spack/cmd/log.py
Comment (scheibelp) [reply lib/spack/spack/cmd/log.py]: (this option has been removed for now, since https://github.com/spack/spack/pull/42202#discussion_r1466230761 means that significantly more logic is required to monitor the introduction of new phase log files)

File: lib/spack/spack/cmd/log.py
Comment (scheibelp) [reply lib/spack/spack/cmd/log.py]: > Also early return missing? 

There was a return one line below this at the time of your review, so I'm not clear what you mean

File: lib/spack/spack/cmd/log.py
Comment (scheibelp) [reply lib/spack/spack/cmd/log.py]: I like the part about opening `stdout` in binary mode (no need for decoding)

I disagree about assuming this is a `.gz` - that seems overly-tied to decisions made outside of this. The current implementation is straightforward as is ("give me the decompressor for the log file" is easier than "this file is obviously a .gz because of code written elsewhere and I'm going to run with that") and also handles the case of older (not-compressed) log files.

File: lib/spack/spack/cmd/log.py
Comment (scheibelp) [reply lib/spack/spack/cmd/log.py]: > which may hit all the bugs already addressed over the decades in coreutils.

Do you have references to these? I'm asking because I'm interested in reading some of this. I couldn't find anything definitive about what is/is-not supported, or about what should/should-not be supported.

Overall this is generally punted on account of your observation in https://github.com/spack/spack/pull/42202#discussion_r1466230761. I'd like to add (in a follow-up PR) a mechanism for auto-detecting the phase logs as they are created and streaming them.

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [inline lib/spack/spack/cmd/logs.py]: I might be able to disable capsys to avoid this in tests. I'm marking this WIP in the meantime. Because this is specifically for dumping content to a terminal, I think we can choose not to support execution of this command in contexts where `sys.stdout` would not be a byte stream.

File: lib/spack/spack/cmd/log.py
Comment (haampie) [reply lib/spack/spack/cmd/log.py]: Why would you choose the more difficult way of doing things if you can just make Spack also write to the the combined logs when it writes to the phase log, and follow that single combined file. If you do it like that you can also delete the code path where logs are combined post install.

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [reply lib/spack/spack/cmd/logs.py]: Split it in two unit tests, and drop the integration test

1. unit test for `dump_logs(input, output)` that accepts an input file name and a writable stream and pass io.BytesIO in the test.
2. command test that mocks `dump_logs` and checks if it's called the right way depending on command line arguments.

File: lib/spack/spack/cmd/log.py
Comment (haampie) [reply lib/spack/spack/cmd/log.py]: If you care so much about future decompression schemes, then branch on the extension: `.gz` for compressed logs, `.txt` / `.log` for uncompressed logs, else error. If somebody would ever add support for .bzip2 / .zstd / whatever, which is never going to happen anyways because Python has no builtin support for it and because there is no signifcant gain in speed / size, your test will error.

I _really_ dislike how you're writing this. I just don't want to see a solution to the trivial "problem" of displaying logs (= dump a file to stdout) that does a double pass over the data, involves creating a temporary directory and a temporary file, as well as chdir twice. If you don't have to store decompressed data, then don't do it, it's 25x the original size.

If you really like to implement it that way then I'll let @tgamblin review it.

File: lib/spack/spack/cmd/log.py
Comment (scheibelp) [reply lib/spack/spack/cmd/log.py]: I'm fine doing it that way instead if it's easier. I don't know if messing with the `log_output` interface is easier, perhaps in this case it wouldn't be so bad, e.g. modifying installer.py/log_output like:

```
    log_contextmanager = log_output(
                        log_file,
                        self.echo,
                        True,
                        env=self.unmodified_env,
                        filter_fn=self.filter_fn,
                        duplicate_to=other_log_file,      # <-------- we are already intercepting everything so we could likely write it twice without much worry
                    )
```

doing it this way makes me wish we weren't compressing the logs in the first place though, since then we could just supply the filenames and let the user `tail -f` it themselves.

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: These are now split up, and I added docs about the expectations for `stdout` when using this command.

File: lib/spack/spack/cmd/log.py
Comment (scheibelp) [reply lib/spack/spack/cmd/log.py]: > If you care so much about future decompression schemes

I don't, and to be clear I explicitly don't care about the *current* decompression scheme, which is why I thought it was nice to shell out to an existing bit of code which handles that for us (the `chdir` stuff is mildly clunky, but completely straightforward, and could have been shunted into `filesystem`). 

> that does a double pass over the data, involves creating a temporary directory and a temporary file, as well as chdir twice. If you don't have to store decompressed data, then don't do it, 

None of these things seem particularly problematic for a command that is explicitly designed to dump everything into the user's shell, which is why I think some pushback was warranted. Since you insist, and I see no problems with doing this, it is now done.

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [inline lib/spack/spack/cmd/logs.py]: Like I suggested before:
```python
with fstream as f:
  _dump_byte_stream_to_stdout(f)
```

No need for a try finally block, which is just a poor man's version of the existing context manager.

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [inline lib/spack/spack/cmd/logs.py]: `extension_from_file` is doing a lot of unnecessary things, and it's a complete misnomer. Should be OK to branch on actual extension `(Pure)Path(p).suffix` or `endswith`.

See also nitpick below about `open`, you shouldn't open or stat the file more than once.

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [inline lib/spack/spack/cmd/logs.py]: Doesn't `os.fdopen(sys.stdout.fileno(), ...)` work everywhere? Looks [it's common enough](https://github.com/search?q=os.fdopen%28sys.stdout.fileno&type=code)

Looks like [pytest itself uses something involved for Windows](https://github.com/pytest-dev/pytest/blob/a164ed6400b675118ba499304ba94a65b8cf9bca/src/_pytest/capture.py#L115-L131), which ultimately actually involves `sys.stdout.fileno`, so I guess it's OK?

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [inline lib/spack/spack/cmd/logs.py]: Nitpick: `os.path.exists` does not imply that `open` works (may no longer exist + permission errors), I would write it "tell-don't-ask":

```python
try:
  handle = gzip.open(log_path) if compressed else open(log_path, "rb")
except OSError as e:
  raise SpackCommandError(f"Cannot read logs for {spec}: {e}") from e
```

Can special case `errno.ENOENT` if you want your original error message, it's mapped correctly on windows.

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: For now this is updated to use `.fileno()` everywhere. 

It seems to work on Windows: I'm less familiar with Windows IO, and given that https://docs.python.org/3/library/sys.html#sys.stdout specifically mentioned using `stdout.buffer`, I thought I'd do it the way they recommend vs. looking up whether a `fileno` would always be available (and whether `fdopen` would always be a reasonable way to interact with it).

https://docs.python.org/3/library/sys.html#sys.stdout does say though that `sys.stdout` is a [file object](https://docs.python.org/3/glossary.html#term-file-object), and on Windows it should have a file handle in all the contexts where it makes sense to invoke `spack logs`.

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: > Nitpick: os.path.exists does not imply that open works (may no longer exist + permission errors),

It is convenient elsewhere in this function to do checks like this to determine whether the package is installed or staged (e.g. `if spec.installed`), so we can already encounter synchronization issues if other processes delete the install prefix or stage; given that, IMO the value of EAFP-style access is reduced. ~I have updated it in this case though.~

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: > it's a complete misnomer

It's definitely not, the test added in this PR (among all the original tests for the function itself) ensure that: I create a `.gz` file, and this function identifies it as a `gz` (otherwise the test would fail trying to read compressed output).

> is doing a lot of unnecessary things,

we have a library function that tells us the extension of a file, I don't think it's the business of this code to worry about the implementation details of that function

>See also nitpick below about `open`, you shouldn't open or stat the file more than once.

As I mention there, this function has no guarantees about the ongoing existence of files (including after we've opened them), so I don't think this is a problem.

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: EDIT: I restored the `if` because it is useful for generating sensible error messages when using `compression.extension_from_file` (which I argue is reasonable here in https://github.com/spack/spack/pull/42202#discussion_r1470093550)

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: > Like I suggested before:

Can you point out where you explicitly suggested this? I just looked through every one of your comments and didn't see it. You do have examples that use a context-manager approach to `open`, but in all cases that I see, they are discussing other things (e.g. [here](https://github.com/spack/spack/pull/42202#discussion_r1466233128) your primary complaint is the use of `decompressor_for`).

That being said, I'm aware of using the context manager approach to opening files, but:

```
if ...:
  fstream = ...
else:
  fstream = ...
<------------ something can happen here, e.g. a signal
with fstream as f:
   ...
```

so I don't consider the context manager sufficient in this case.

Again, if you explicitly requested using the context manager approach in an earlier comment, then I should have explained my reasoning there, and I apologize for not doing so until now.

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [reply lib/spack/spack/cmd/logs.py]: Can you please look at the implementation? :)

It's not deriving an "extension" from a file name, it classifies the content type from magic bytes. So it opens the file. You want to branch on extension here, cause otherwise you're interacting with the file on the file system yet a third time, which can raise exceptions, which you don't handle.

> worry about the implementation details

If you expect it cannot raise exceptions, maybe. It's worth not treating other code as a black box for error handling.

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [reply lib/spack/spack/cmd/logs.py]: https://github.com/spack/spack/pull/42202#discussion_r1466233128

```<------------ something can happen here, e.g. a signal```

I doubt it makes sense to care. The only difference is assignment to a variable. Why would using the return value from a function immediately in a `with open(...)` block make any difference to `x = open(...); with x`

What do you think about the code that runs between when gzip.open opens the file and before it returns? I think if you're truly worried about this you should open the file in binary mode in a context manager before wrapping it in a gzip.GzipFile, but I don't think that is worth doing. If the process exits over an interrupt or whatever, then python garbage collection closes the file, and otherwise the system.

You can use the context manager, it's fine, and it's much cleaner.

File: lib/spack/spack/cmd/logs.py
Comment (haampie) [reply lib/spack/spack/cmd/logs.py]: Sure. I guess for `spec.installed` you can prolong the read lock, but also I wouldn't care _that_ much.

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: > #42202 (comment)

If this is meant to be the example of you being explicit, I can say that comment is the exact example I gave of you not being explicit

> What do you think about the code that runs between when gzip.open opens the file and before it returns

That's up to python's implementation of context managers, and `gzip`'s implementation of `.open` as a context manager: I can  trust `with gzip.open as f` the same way I trust `with open() as f`.

> if you're truly worried about this you should open the file in binary mode in a context manager before wrapping it in a gzip.GZipFile, but I don't think that is worth doing.

I addressed the concern I had in the way that was natural to me, this notion of `finally` closing a file being messy seems extremely odd to me: reading into the interface of `gzip` to understand that it can wrap an already open stream is perfectly fine, but I don't see how it's any simpler than realizing we need to `close` our open file.

Now that you've pointed it out, I can change it, but my point is that holding up the PR on something like this seems to be emphasizing the wrong thing.

File: lib/spack/spack/cmd/logs.py
Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: > It's not deriving an "extension" from a file name,

I never said it did. My point is that I don't care how it derives the extension: I give it a file name, it gives me the extension.

> cause otherwise you're interacting with the file on the file system yet a third time, which can raise exceptions, which you don't handle.

A recurring theme of your comments seems to be a concern about the possibility of the state of the file changing between my checks. This command has no control over that because the files in question can be deleted by another process at any time. I don't handle those exceptions because there is no recourse for handling them and they should in fact terminate the process.

Also, the fact that the file might be deleted in this scenario

```
ext = compression.extension_from_file(path)
# the file was deleted
if ext == .gz
  gzip.open
```

is the same as this scenario

```
ext = path.split()[-1] # or whatever other non-fs-interacting mechanism you intend to use to determine the extension
# The file was deleted right here
if ext == .gz
  gzip.open
```

I don't care in the first case as much as I don't care in the second case: I don't depend on the continued existence of the file between determining the extension and trying to open it.

Comment (haampie) [reply lib/spack/spack/cmd/logs.py]: It is not the same scenario if you take my advice into account.

```python
try:
  handle = open(path, "rb")
except OSError as e:
  # deal with non-existence / permission errors once and only once.
  raise ... from e
with handle as f:
  do_something(maybe_decompress(f))
```

there should not be stat or open calls above `try`, so no `extension_from_file` calls, opening once is enough.

My point is rather that you seem to care about "nice" error messages initially, but right after forget to deal with exceptions from further `open` calls.

If you care then why don't you make an effort to make those impossible.

Comment (scheibelp) [reply lib/spack/spack/cmd/logs.py]: One new implied desire in https://github.com/spack/spack/pull/42202#discussion_r1470232557 is for better error messages. Up to this point, my code had no `except` in it, so I will argue that this is all effectively the same in that case. However, if `except` is used, then paying attention to the possible errors of `compression.extension_from_file` makes sense: [4d44aec](https://github.com/spack/spack/pull/42202/commits/4d44aec8361650cbbd9fd7d595224e72fe27f97a) achieves that by wrapping the extension detection and `open()` into a single `try` block (and updates the tests for that).

**Review by haampie — APPROVED** [review]

**Pull request merged**
